File: python3/practice/classic_puzzle/easy/six-degrees-of-kevin-bacon.py
# import sys
# import math

import itertools

# 6 Degrees of Kevin Bacon!

actor_name = input()
n = int(input())
movie_casts = []
for _ in range(n):
    movie_cast = input()
    movie_casts.append(movie_cast)

# Write an answer using print
# To debug: print("Debug messages...", file=sys.stderr, flush=True)

kevin = 'Kevin Bacon'

casts_list = []
for movie_cast in movie_casts:
    _movie_name, movie_casts = movie_cast.split(': ', 1)
    casts_list.append(movie_casts.split(', '))

target_casts = [actor_name]
bacon_number = 0
while kevin not in target_casts:
    bacon_number += 1

    next_target_casts = itertools.chain.from_iterable([
        casts
        for casts
        in casts_list
        if any(target in casts for target in target_casts)
    ])

    target_casts = list(set(next_target_casts))

result = bacon_number

# print("N degrees to Kevin Bacon")
print(result)

File: typescript/practice/classic_puzzle/easy/six-degrees-of-kevin-bacon.ts
/**
 * 6 Degrees of Kevin Bacon!
 **/

const actorName: string = readline();
const n: number = parseInt(readline());
const movieCasts: string[] = [];
for (let i = 0; i < n; i++) {
    const movieCast: string = readline();
    movieCasts.push(movieCast);
}

// Write an answer using console.log()
// To debug: console.error('Debug messages...');

const kevin: string = 'Kevin Bacon';

const castsArray: string[][] = movieCasts.map(movieCast => {
    const [_movieName, casts]: string[] = movieCast.split(': ');

    return casts.split(', ');
});

let targetCasts: string[] = [actorName];
let baconNumber: number = 0;
while (!targetCasts.includes(kevin)) {
    baconNumber++;

    const nextTargetCasts = castsArray.flatMap(casts => {
        const isFoundTargetCast: boolean = targetCasts.some(target => casts.includes(target));

        return isFoundTargetCast ? casts : [];
    });

    targetCasts = nextTargetCasts;
}

const result: string = baconNumber.toString();

// console.log('N degrees to Kevin Bacon');
console.log(result);
